design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,509 @@
+the Certificate resource, this is a logical set of policy options the cluster
+operator can configure. Any extension to this base set of rules can be
+implemented using an external policy engine.
+
+Apart from verifying external policy servers, fields are validated at admission
+time. The status condition is primarily used for exposing the status of the
+external policy servers, which are periodically checked.
+
+```go
+type CertificateRequestPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec CertificatePolicySpec `json:"spec"`
+
+    // +optional
+    Status CertificatePolicyStatus `json:"status"`
+}
+
+type CertificateRequestPolicyList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+
+    Items []Policy `json:"items"`
+}
+
+type CertificateRequestPolicySpec struct {
+    // +optional
+    AllowedSubject *PolicyX509Subject `json:"allowedSubject,omitempty"`
+
+    // +optional
+    AllowedCommonName *string `json:"subject,omitempty"`
+
+    // Values are inclusive (i.e. a min value with 50s will accept a duration
+    // with 50s). MinDuration and MaxDuration may be the same.
+    // +optional
+    MinDuration *metav1.Duration `json:"minDuration,omitempty"`
+    // +optional
+    MaxDuration *metav1.Duration `json:"maxDuration,omitempty"`
+
+    // +optional
+    AllowedDNSNames *[]string `json:"allowedDNSNames,omitempty"`
+
+    // +optional
+    AllowedIPAddresses *[]string `json:"allowedIPAddresses,omitempty"`
+
+    // +optional
+    AllowedURIs *[]string `json:"allowedURIs,omitempty"`
+
+    // +optional
+    AllowedEmailAddresses *[]string `json:"allowedEmailAddresses,omitempty"`
+
+    // +optional
+    AllowedIssuers *[]cmmeta.ObjectReference `json:"allowedIssuer,omitempty"`
+
+    // +optional
+    AllowedIsCA *bool `json:"allowedIsCA,omitempty"`
+
+    // +optional
+    AllowedUsages *[]cmapi.KeyUsage `json:"allowedUsages,omitempty"`
+
+    // +optional
+    AllowedPrivateKey *PolicyPrivateKey `json:"allowedPrivateKey,omitempty"`
+
+    // +optional
+    ExternalPolicyServers []string `json:"externalPolicyServers,omitempty"`
+}
+
+type PolicyX509Subject struct {
+    // +optional
+    AllowedOrganizations *[]string `json:"allowedOrganizations,omitempty"`
+    // +optional
+    AllowedCountries *[]string `json:"allowedCountries,omitempty"`
+    // +optional
+    AllowedOrganizationalUnits *[]string
+    // +optional
+    AllowedLocalities *[]string `json:"allowedLocalities,omitempty"`
+    // +optional
+    AllowedProvinces *[]string `json:"allowedProvinces,omitempty"`
+    // +optional
+    AllowedStreetAddresses *[]string `json:"allowedStreetAddresses,omitempty"`
+    // +optional
+    AllowedPostalCodes *[]string `json:"allowedPostalCodes,omitempty"`
+    // +optional
+    AllowedSerialNumber *string `json:"allowedSerialNumber,omitempty"`
+}
+
+type CertificateRequestPolicyPrivateKey struct {
```

Comment:
Nitpick: Should this be "CertificateRequestKeyPolicy" rather than "CertificateRequestPolicyPrivateKey"?

e.g. `MinSize` and `MaxSize` refer to the length of the RSA key modulus, which is not specific to the private key e.g. https://security.stackexchange.com/a/90182

That said, I see that there are existing public types used (e.g. `cmapi.PrivateKeyAlgorithm`) in cert-manager which use this naming convention, so maybe it's better to preserve the naming convention already used elsewhere